apm-agent-premain/src/main/java/co/elastic/apm/agent/premain/AgentMain.java

```diff
@@ -101,13 +99,72 @@ public synchronized static void init(String agentArguments, Instrumentation inst
             if (doDisable) {
                 return;
             }
+        }
 
+        // workaround for classloader deadlock https://bugs.openjdk.java.net/browse/JDK-8194653
+        FileSystems.getDefault();
+
+        long delayAgentInitMs = -1L;
+        String delayAgentInitMsProperty = System.getProperty("elastic.apm.delay_agent_premain_ms");
+        if (delayAgentInitMsProperty != null) {
+            try {
+                delayAgentInitMs = Long.parseLong(delayAgentInitMsProperty.trim());
+            } catch (NumberFormatException numberFormatException) {
+                System.err.println("The value of the \"elastic.apm.delay_agent_premain_ms\" System property must be a number");
+            }
         }
+        if (premain && shouldDelayOnPremain()) {
+            delayAgentInitMs = Math.max(delayAgentInitMs, 3000L);
+        }
+        if (delayAgentInitMs > 0) {
+            delayAndInitAgentAsync(agentArguments, instrumentation, premain, delayAgentInitMs);
+        } else {
+            loadAndInitializeAgent(agentArguments, instrumentation, premain);
+        }
+    }
 
-        try {
-            // workaround for classloader deadlock https://bugs.openjdk.java.net/browse/JDK-8194653
-            FileSystems.getDefault();
+    /**
+     * Returns whether agent initialization should be delayed when occurring through the {@code premain} route.
+     * This works around a JVM bug (https://bugs.openjdk.java.net/browse/JDK-8041920) causing JIT fatal error if
+     * agent code causes the loading of MethodHandles prior to JIT compiler initialization.
+     * @return {@code true} for any Java 7 and early Java 8 HotSpot JVMs, {@code false} for all others
```

Comment:
Removing that constraint on early Java 8 early versions would be really nice !! Do we need to make load-testing able to run tests with forks to validate this ? We could use a temporary branch on the project itself as a work-around.

Reply:
Indeed, I thought saw to, but @felixbarny pointed out to me that we already had this restriction before moving to `invokedynamic` (although we used MethodHandles before). I think that using so old Java 8 updates is less common than using Java 7, so let's keep this restriction and this handles cases where some IS trying to use our agent on these old JVMs and disables the bootstrap checks